runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainer.java

```diff
@@ -0,0 +1,300 @@
+
+
+/**
+ * Spark accumulator value which holds all {@link MetricsContainer}s, aggregates and merges them.
+ */
+public class SparkMetricsContainer implements Serializable {
+  private static final Logger LOG = LoggerFactory.getLogger(SparkMetricsContainer.class);
+
+  private transient volatile LoadingCache<String, MetricsContainer> metricsContainers;
+
+  private final Map<MetricKey, CounterAggregator> counters = new HashMap<>();
+  private final Map<MetricKey, DistributionAggregator> distributions = new HashMap<>();
+
+  SparkMetricsContainer() {}
+
+  public static Accumulator<SparkMetricsContainer> getAccumulator(JavaSparkContext jsc) {
+    return MetricsAccumulator.getInstance(jsc);
+  }
+
+  public MetricsContainer getContainer(String stepName) {
+    if (metricsContainers == null) {
+      synchronized (this) {
+        if (metricsContainers == null) {
+          metricsContainers = CacheBuilder.<String, SparkMetricsContainer>newBuilder()
+              .build(new MetricsContainerCacheLoader());
+        }
+      }
+    }
+    try {
+      return metricsContainers.get(stepName);
+    } catch (ExecutionException e) {
+      LOG.error("Error while creating metrics container", e);
+      return null;
+    }
+  }
+
+  Collection<CounterAggregator> getCounters() {
+    return counters.values();
+  }
+
+  Collection<DistributionAggregator> getDistributions() {
+    return distributions.values();
+  }
+
+  SparkMetricsContainer merge(SparkMetricsContainer other) {
+    SparkMetricsContainer merged = new SparkMetricsContainer();
+    merged.aggregate(this.counters.values(), this.distributions.values());
+    merged.aggregate(other.counters.values(), other.distributions.values());
+    return merged;
+  }
+
+  private void writeObject (ObjectOutputStream out) throws IOException {
+    aggregate();
+    out.defaultWriteObject();
+  }
+
+  private void aggregate() {
+    if (metricsContainers != null) {
+      for (MetricsContainer container : metricsContainers.asMap().values()) {
+        MetricUpdates cumulative = container.getCumulative();
+        Iterable<CounterAggregator> counterAggregators =
+            Iterables.transform(cumulative.counterUpdates(),
+                new Function<MetricUpdate<Long>, CounterAggregator>() {
+                  @SuppressWarnings("ConstantConditions")
+                  @Nullable
+                  @Override
+                  public CounterAggregator
+                  apply(MetricUpdate<Long> update) {
+                    return new CounterAggregator(new SparkMetricKey(update.getKey()),
+                        update.getUpdate());
+                  }
+                });
+        Iterable<DistributionAggregator> distributionAggregators =
+            Iterables.transform(cumulative.distributionUpdates(),
+                new Function<MetricUpdate<DistributionData>, DistributionAggregator>() {
+                  @SuppressWarnings("ConstantConditions")
+                  @Nullable
```

Comment:
I don't think `@Nullable` is required in cases like this where the implementation is `return new ...`